File: config/mapping.py
from imposm.mapping import (
    Options,
    Points, LineStrings, Polygons,
    String, Bool, Integer, OneOfInt, FieldType,
    set_default_name_type, LocalizedName,
    WayZOrder, ZOrder, Direction,
    GeneralizedTable, UnionView,
    PseudoArea, meter_to_mapunit, sqr_meter_to_mapunit,
)

import yaml
import os


db_yml = yaml.load(file(os.getenv("DB_CONFIG") or "config/database.yml"))
db_env = db_yml[os.getenv("RAILS_ENV") or "development"]

db_conf = Options(
    host=db_env['host'],
    user=db_env['username'],
    password=db_env['password'],
    port=5432,
    sslmode='allow',
    prefix='osm_new_',
    proj='epsg:900913',
)

set_default_name_type(LocalizedName(['name:ru','name','int_name']))

class Length(FieldType):
    column_type = "REAL"

    def value(self, val, osm_elem):
        length = osm_elem.geom.length
        if not length:
            return None

        return length * (40075160 / 360)

    def extra_fields(self):
        return []

class NotArea(Bool):

    def value(self, val, osm_elem):
        if hasattr(osm_elem.coords, 'type'): # Ugly way to detect relations
            return 0

        if val is None or val.strip().lower() in self.neg_aliases:
            return 1  # not self.default
        return 0  # self.default

places = Points(
    name = 'places',
    mapping = {
        'place': (
            'city',
            'town',
            'village'
        ),
    },
    fields = (
        ('z_order', ZOrder(['city','town','village'])),
        ('population', Integer()),
    )
)

roads = LineStrings(
    name = 'roads',
    mapping = {
        'highway': ('__any__',),
    },
    fields = (
        ('z_order', WayZOrder()),
        ('tunnel', Bool()),
        ('bridge', Bool()),
        ('covered', Bool()),
        ('surface', String()),
        ('ref', String()),
        ('length', Length()),
    ),
    field_filter = (
        ('area', Bool()),
    )
)

railways = LineStrings(
    name = 'railways',
    mapping = {
        'railway': ('__any__',),
    },
    fields = (
        ('z_order', WayZOrder()),
        ('tunnel', Bool()),
        ('bridge', Bool()),
        ('length', Length()),
    )
)

buildings = Polygons(
    name = 'buildings',
    mapping = {
        'building': ('__any__',),
    },
    fields = (
        ('addr:housenumber', String()),
    )
)

waterways = LineStrings(
    name = 'waterways',
    mapping = {
        'barrier': (
            'ditch'
        ),
        'waterway': (
            'stream',
            'river',
            'canal',
            'drain',
            'ditch'
        ),
    },
    fields = (
        ('length', Length()),
    ),
    field_filter = (
        ('tunnel', Bool()),
    ),
)

waterareas = Polygons(
    name = 'waterareas',
    mapping = {
        'waterway': (
            'riverbank',
            'dock',
            'mill_pond',
            'canal'
        ),
        'natural': (
            'water',
            'lake',
            'bay'
        ),
        'landuse': (
            'basin',
            'reservoir'
        ),
    },
    fields = (
        ('area', PseudoArea()),
    )
)

landuses = Polygons(
    name = 'landuses',
    mapping = {
        'landuse': ('__any__',),
    },
    fields = (
        ('area', PseudoArea()),
        ('z_order', ZOrder([
            'construction',
            'cemetery',
            'residential',
            'industrial',
            'commercial',
        ])),
    )
)

squares = Polygons(
    name = 'squares',
    mapping = {
        'highway': ('__any__',),
    },
    field_filter = (
        ('area', NotArea()),
    )
)

areas = Polygons(
    name = 'areas',
    mapping = {
        'leisure': (
            'park',
            'garden',
            'playground',
            'golf_course',
            'sports_centre',
            'pitch',
            'stadium',
            'common',
            'nature_reserve',
        ),
        'natural': (
            'wood',
            'land',
            'scrub',
            'wetland',
            'beach',
        ),
        'amenity': ('__any__',),
        'tourism': ('__any__',),
    },
    fields = (
        ('area', PseudoArea()),
        ('z_order', ZOrder([
            'fuel',
            'zoo',
            'footway',
            'pedestrian',
            'parking',
            'school',
            'university',
            'college',
            'hospital',
            'place_of_worship',
            'sports_centre',
            'stadium',
            'pitch',
            'garden',
            'park',
            'playground',
            'common',
            'wood',
            'scrub',
            'wetland',
            'beach',
        ])),
    ),
    field_filter = (
        ('building', Bool()),
    )
)

barrier_lines = LineStrings(
    name = 'barrier_lines',
    mapping = {
        'barrier': (
            'city_wall',
            'fence',
            'hedge',
            'retaining_wall',
            'wall',
            'bollard',
            'gate',
            'spikes',
            'lift_gate',
            'kissing_gate',
            'embankment',
            'yes',
            'wire_fence',
        ),
        'natural': (
            'cliff',
        ),
        'man_made': (
            'enbankment',
        )
    },
    fields = (
        ('length', Length()),
    )
)

barrier_polys = Polygons(
    name = 'barrier_polys',
    mapping = {
        'barrier': (
            'city_wall',
            'fence',
            'hedge',
            'retaining_wall',
            'wall',
            'bollard',
            'gate',
            'spikes',
            'lift_gate',
            'kissing_gate',
            'embankment',
            'yes',
            'wire_fence',
        ),
        'natural': (
            'cliff',
        ),
        'man_made': (
            'enbankment',
        )
    },
    fields = (
        ('length', Length()),
    )
)

barriers = UnionView(
    name = 'barriers',
    fields = (),
    mappings = [barrier_lines, barrier_polys],
)

roads_gen = GeneralizedTable(
    name = 'roads_gen',
    tolerance = meter_to_mapunit(50.0),
    origin = roads,
)

railways_gen = GeneralizedTable(
    name = 'railways_gen',
    tolerance = meter_to_mapunit(50.0),
    origin = railways,
)

File: config/mapping.rb
multipolygons :places do
  with :center

  map :place => [:city, :town, :village]

  after_import do
    conn.exec "INSERT INTO #{name}(id,osm_type,type,name,center,tags) SELECT #{osm_id_expr}, #{osm_type_expr :point}, #{type_mapping}, #{name_mapping}, way, tags FROM raw_osm_point src WHERE #{conditions} AND (SELECT COUNT(1) FROM #{name} ep WHERE ep.geometry && src.way AND ST_Contains(Geometry(ep.geometry), src.way)) = 0"
    conn.exec "UPDATE #{name} SET tags = #{name}.tags || src.tags, center = src.way FROM raw_osm_point src WHERE #{conditions} AND #{name_mapping} = #{name}.name AND geometry && way AND ST_Contains(Geometry(geometry), way)"
  end
end

lines :roads do
  with :ref => :string, :place_id => :id

  map :highway

  after_import do
    conn.exec "UPDATE #{name} SET place_id = new_osm_places.id from new_osm_places where st_intersects(new_osm_places.geometry,new_osm_roads.geometry) and new_osm_places.name is not null"
  end
end

multilines :streets do
  with :center, :city => :string, :place_id => :id

  after_create do # TODO More intellectual type selection
    conn.exec "INSERT INTO #{name}(id,osm_type,type,geometry,name,place_id) SELECT MIN(id), MIN(osm_type), 'street', ST_Multi(ST_Union(Geometry(geometry))), name, place_id FROM new_osm_roads WHERE name IS NOT NULL AND place_id IS NOT NULL GROUP BY name, place_id"
    conn.exec "UPDATE #{name} SET city = (SELECT name FROM new_osm_places WHERE id = place_id)"
  end
end

lines :rails do
  map :railway
end

polygons :squares do
  map :highway
end

lines :waterways do
  map :waterway => [:stream, :river, :canal, :drain, :ditch], :barrier => :ditch
end

multipolygons :waterareas do
  map :waterway => [:riverbank, :drain, :pond], :natural => [:water, :lake, :bay], :landuse => [:basin, :reservoir]
end

multilines :barriers do
  map :barrier, :natural => :cliff, :man_made => :enbankment

  after_import do
    conn.exec "INSERT INTO #{name}(id,osm_type,type,geometry) SELECT #{osm_id_expr}, #{osm_type_expr :polygon}, #{type_mapping} AS type, ST_Multi(ST_Boundary(way)) FROM raw_osm_polygon src WHERE #{conditions}"
  end
end

multipolygons :territories do
  map :landuse, :natural => [:wood, :scrub, :wetland, :beach]
end

multipolygons :buildings do
  map :building, :power => :generator, :man_made => [:water_tower, :reservoir_covered, :tank, :water_tank, :water_works, :wastewater_plant, :tower, :communications_tower, :monitoring_station]

  with :address
end

multipolygons :objects do
  map :leisure, :amenity, :tourism, :shop, :office, :sport, :man_made => [:well, :water_well, :artesian_well], :historic => [:monument, :memorial, :ruins], :landuse => [:cemetery], :natural => [:spring], :multi => true

  with :address, :center

  after_import do
    conn.exec "INSERT INTO #{name}(id,osm_type,type,types,name,center,tags) SELECT #{osm_id_expr}, #{osm_type_expr :point}, #{type_mapping}, #{types_mapping}, #{name_mapping}, way, tags FROM raw_osm_point src WHERE #{conditions}"
    conn.exec "UPDATE #{name} SET city = COALESCE(#{name}.city,b.city), street = COALESCE(#{name}.street, b.street), housenumber = COALESCE(#{name}.housenumber, b.housenumber), postcode = COALESCE(#{name}.postcode, b.postcode) FROM new_osm_buildings b WHERE ST_Intersects(b.geometry,new_osm_objects.geometry) OR ST_Intersects(b.geometry,new_osm_objects.center)"
    conn.exec "UPDATE #{name} SET city = p.name FROM new_osm_places p WHERE #{name}.geometry && p.geometry AND ST_Contains(Geometry(p.geometry), Geometry(#{name}.geometry)) AND p.type IN ('city','town','village') AND p.name IS NOT NULL AND city IS NULL AND #{name}.geometry IS NOT NULL"
    conn.exec "UPDATE #{name} SET city = p.name FROM new_osm_places p WHERE #{name}.center && p.geometry AND ST_Contains(Geometry(p.geometry), Geometry(#{name}.center)) AND p.type IN ('city','town','village') AND p.name IS NOT NULL AND city IS NULL AND #{name}.geometry IS NULL"

    conn.exec "UPDATE new_osm_buildings SET name = NULL FROM #{name} WHERE #{name}.id = new_osm_buildings.id" # Remove names from buildings to exclude them from search
  end
end
